File: src/neml2/dispatchers/StaticHybridScheduler.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include <set>

#include "neml2/dispatchers/StaticHybridScheduler.h"
#include "neml2/misc/assertions.h"
#include "neml2/base/Registry.h"

namespace neml2
{
register_NEML2_object(StaticHybridScheduler);

OptionSet
StaticHybridScheduler::expected_options()
{
  OptionSet options = WorkScheduler::expected_options();
  options.doc() = "Dispatch work to multiple devices based on provided batch sizes and priorities.";

  options.add<std::vector<Device>>("devices", "List of devices to dispatch work to");
  options.add<std::vector<std::size_t>>("batch_sizes", "List of batch sizes for each device");
  options.add_optional<std::vector<std::size_t>>(
      "capacities", "List of capacities for each device, default to batch_sizes");
  options.add_optional<std::vector<double>>("priorities", "List of priorities for each device");

  return options;
}

StaticHybridScheduler::StaticHybridScheduler(const OptionSet & options)
  : WorkScheduler(options),
    _available_devices(options.get<std::vector<Device>>("devices"))
{
}

void
StaticHybridScheduler::setup()
{
  WorkScheduler::setup();

  const auto & device_list = input_options().get<std::vector<Device>>("devices");
  const auto & batch_sizes = input_options().get<std::vector<std::size_t>>("batch_sizes");
  const auto & capacities = input_options().defined("capacities")
                                ? input_options().get<std::vector<std::size_t>>("capacities")
                                : std::vector<std::size_t>{};
  const auto & priorities = input_options().defined("priorities")
                                ? input_options().get<std::vector<double>>("priorities")
                                : std::vector<double>{};

  // First pass:
  // - Check if any CPU device is present
  // - Check if any CUDA device is present
  // - Make sure no more than one CPU device is present
  for (const auto & device : device_list)
    if (device.is_cpu())
    {
      neml_assert(!_cpu, "Multiple CPU devices are not allowed");
      _cpu = true;
    }
    else if (device.is_cuda())
      _cuda = true;
    else
      neml_assert(false, "Unsupported device type: ", device);

  // Second pass:
  // - If multiple CUDA devices are present, make sure each CUDA device has a concrete
  //   (nonnegative), unique device ID
  bool has_multiple_cuda_devices = _cpu ? device_list.size() > 2 : device_list.size() > 1;
  if (has_multiple_cuda_devices)
  {
    std::set<DeviceIndex> cuda_device_ids;
    for (const auto & device : device_list)
    {
      if (device.is_cpu())
        continue;
      auto device_id = device.index();
      neml_assert(device_id >= 0, "Device ID must be nonnegative");
      neml_assert(cuda_device_ids.find(device_id) == cuda_device_ids.end(),
                  "Device ID must be unique. Found duplicate: ",
                  device_id);
      cuda_device_ids.insert(device_id);
    }
  }

  // Expand batch size if necessary
  auto batch_sizes_expand = batch_sizes;
  if (batch_sizes.size() == 1)
    batch_sizes_expand.resize(device_list.size(), batch_sizes[0]);
  else
    neml_assert(batch_sizes.size() == device_list.size(),
                "Number of batch sizes must either be one or match the number of devices.");

  // Expand capacity if necessary
  auto capacities_expand = capacities;
  if (capacities.empty())
    capacities_expand = batch_sizes_expand;
  else if (capacities.size() == 1)
    capacities_expand.resize(device_list.size(), capacities[0]);
  else
    neml_assert(capacities.size() == device_list.size(),
                "Number of capacities must either be zero, one, or match the number of devices.");

  // Expand priorities if necessary
  auto priorities_expand = priorities;
  if (priorities.empty())
    priorities_expand.resize(device_list.size(), 1.0);
  else
    neml_assert(priorities.size() == device_list.size(),
                "Number of priorities must match the number of devices.");

  // Construct the device status list
  for (std::size_t i = 0; i < device_list.size(); ++i)
    _devices.emplace_back(
        device_list[i], batch_sizes_expand[i], capacities_expand[i], priorities_expand[i]);
}

void
StaticHybridScheduler::set_availability_calculator(std::function<double(const DeviceStatus &)> f)
{
  _custom_availability_calculator = std::move(f);
}

bool
StaticHybridScheduler::schedule_work_impl(Device & device, std::size_t & n) const
{
  bool available = false;
  double max_availability = std::numeric_limits<double>::lowest();

  for (const auto & i : _devices)
    if ((i.load + i.batch_size) <= i.capacity)
    {
      auto availability =
          _custom_availability_calculator ? _custom_availability_calculator(i) : i.priority;
      if (!available || availability > max_availability)
      {
        available = true;
        device = i.device;
        n = i.batch_size;
      }
    }

  return available;
}

void
StaticHybridScheduler::dispatched_work_impl(Device device, std::size_t n)
{
  for (auto & i : _devices)
    if (i.device == device)
    {
      i.load += n;
      // TODO: Add an option to allow for oversubscription, maybe?
      neml_assert(i.load <= i.capacity, "Device oversubscribed");
      return;
    }

  neml_assert(false, "Device not found: ", device);
}

void
StaticHybridScheduler::completed_work_impl(Device device, std::size_t n)
{
  for (auto & i : _devices)
    if (i.device == device)
    {
      neml_assert(i.load >= n, "Device load underflow");
      i.load -= n;
      return;
    }

  neml_assert(false, "Device not found: ", device);
}

bool
StaticHybridScheduler::all_work_completed() const
{
  for (const auto & i : _devices)
    if (i.load > 0)
      return false;
  return true;
}
} // namespace neml2

File: python/static-hybrid-scheduler.js
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

class StaticHybridScheduler {
  constructor(workdispatcher) {
    this.work = workdispatcher.works[0];
    this.work_size = workdispatcher.work_sizes[0];
    this.devices = workdispatcher.devices;
    this.batch_sizes = workdispatcher.batch_sizes;
    this.capacities = workdispatcher.capacities;
    this.speeds = workdispatcher.speeds;
  }

  apply_defaults(options) {
    options.dispatch_speed ||= 500; // ms
    options.retrieve_speed ||= 500; // ms
  }

  // Find the next device with the least load
  next(loads) {
    const load_fractions = this.capacities.map((cap, i) =>
      Math.ceil((loads[i] / cap) * 100)
    );
    const min_load = Math.min(...load_fractions);
    return load_fractions.indexOf(min_load);
  }

  // Simulate the work dispatching process
  simulate(options = {}) {
    this.apply_defaults(options);

    let work_events = {};
    let queues = Array.from({ length: this.devices.length }, () => []);
    let time = 0;
    let loads = Array(this.devices.length).fill(0);
    let dev = this.next(loads);
    let queue = queues[dev];
    let next_work_id = 0;
    let next_batch_size = Math.min(this.batch_sizes[dev], this.work_size);
    while (this.work_size > 0) {
      // Dispatch if we can
      if (this.capacities[dev] - loads[dev] >= next_batch_size) {
        let queue = queues[dev];
        // Dispatch a batch of work to the device
        let work_event = {
          size: next_batch_size,
          dispatched_to: dev,
          dispatched_at: time,
          shifted_at: [],
          shifted_by: [],
          process_duration: Math.ceil(next_batch_size / this.speeds[dev]),
        };
        work_event.processed_at = time + options.dispatch_speed;
        work_event.retrieved_at =
          work_event.processed_at + work_event.process_duration;
        this.enqueue(queue, next_work_id, work_event, work_events, options);
        work_events[next_work_id] = work_event;
        // Finish dispatching
        time += options.dispatch_speed;
        this.work_size -= next_batch_size;
        loads[dev] += next_batch_size;
        dev = this.next(loads);
        queue = queues[dev];
        next_work_id++;
        next_batch_size = Math.min(this.batch_sizes[dev], this.work_size);
      } else {
        time = Number.MAX_SAFE_INTEGER;
        for (let next_dev = 0; next_dev < this.devices.length; next_dev++) {
          if (queues[next_dev].length > 0) {
            const first_in_queue = work_events[queues[next_dev][0]];
            const next_time =
              first_in_queue.retrieved_at + options.retrieve_speed;
            if (next_time < time) {
              time = next_time;
              dev = next_dev;
              queue = queues[dev];
            }
          }
        }
        const retrieved = this.dequeue(queue, work_events, options);
        loads[dev] -= retrieved.size;
        dev = this.next(loads);
        next_batch_size = Math.min(this.batch_sizes[dev], this.work_size);
      }
    }

    // Finish queues
    queues.forEach((queue) => {
      while (queue.length > 0) {
        this.dequeue(queue, work_events, options);
      }
    });

    this.consolidate_shifts(work_events);
    return [work_events];
  }

  dequeue(queue, work_events, options) {
    const first_in_queue = work_events[queue.shift()];
    queue.forEach((id) => {
      if (work_events[id].dispatched_at < first_in_queue.dispatched_at) return;
      const shifted_at = Math.max(
        first_in_queue.retrieved_at,
        work_events[id].dispatched_at + options.dispatch_speed
      );
      work_events[id].shifted_at.push(shifted_at);
      work_events[id].shifted_by.push(first_in_queue.size);
    });
    return first_in_queue;
  }

  enqueue(queue, id, next_work_event, work_events) {
    next_work_event.pos0 = 0;
    queue.forEach((i) => {
      next_work_event.pos0 += work_events[i].size;
    });
    if (queue.length === 0) {
      queue.push(id);
      return;
    }
    for (let i = 0; i < queue.length; i++) {
      const work_event = work_events[queue[i]];
      if (next_work_event.retrieved_at < work_event.retrieved_at) {
        queue.splice(i, 0, id);
        return;
      }
    }
    queue.push(id);
  }

  consolidate_shifts(work_events) {
    Object.values(work_events).forEach((work_event) => {
      let shift_by = {};
      work_event.shifted_at.forEach((shifted_at, j) => {
        if (shifted_at in shift_by) {
          shift_by[shifted_at] += work_event.shifted_by[j];
        } else shift_by[shifted_at] = work_event.shifted_by[j];
      });
      work_event.shifted_at = Object.keys(shift_by).map((k) => parseInt(k));
      work_event.shifted_by = Object.values(shift_by);
    });
  }
}
